src/librustc_interface/util.rs

```diff
@@ -741,7 +745,11 @@ impl<'a> MutVisitor for ReplaceBodyWithLoop<'a, '_> {
                 for s in b.stmts {
                     let old_blocks = self.nested_blocks.replace(vec![]);
 
-                    stmts.extend(self.flat_map_stmt(s).into_iter().filter(|s| s.is_item()));
+                    stmts.extend(self.flat_map_stmt(s).into_iter().filter(|s| {
+                        s.is_item()
+                            || s.is_expr_with(expr_is_closure)
+                            || s.is_semi_expr_with(expr_is_closure)
```

Comment:
You should use a single  `fn expr(&self) -> Option<&ast::Expr>` meth `ast::Stmt` instead. I think it's very rare that you would want to call `is_expr_with` without `is_semi_expr_with`.